### PR TITLE
Fixing NSFS semaphore leak

### PR DESCRIPTION
**Pull request by jeniawhite**
### Explain the changes
1. Handling the promises in the pipeline differently (caused unsettled promises on rejections)
2. Rewrite transform and flush async logic with the addition of try and catch to avoid unhandled promise rejections

### Issues: Fixed #xxx / Gap #xxx
1. https://github.com/noobaa/noobaa-core/issues/6848

### Testing Instructions:
1.

### CODE REVIEW

File: src/util/chunk_fs.js [hunk 1]
```diff
@@ -28,54 +28,50 @@ class ChunkFS extends stream.Transform {
...
+                size: chunk.length,
+                count: this.count
+            });
+            this.count = 0;
+            while (chunk && chunk.length) {
+                const available_size = config.NSFS_BUF_SIZE - this.q_size;
+                const buf = (available_size < chunk.length) ? chunk.slice(0, available_size) : chunk;
+                this.q_buffers.push(buf);
+                this.q_size += buf.length;
+                if (this.q_size === config.NSFS_BUF_SIZE) await this._flush_buffers();
```
Comment (guymguym) [inline src/util/chunk_fs.js:45]: I'd use >= just being defensive
```suggestion
                if (this.q_size >= config.NSFS_BUF_SIZE) await this._flush_buffers();
```

**Review by guymguym — COMMENTED** [review]
Missing details on what is the fix itself, and what is the pitfall to avoid in such a code. I'd put it in code comments or in the PR main comment, because I already forgot what was the exact fix here that we discussed before...

File: src/util/chunk_fs.js [hunk 2]
```diff
@@ -28,54 +28,50 @@ class ChunkFS extends stream.Transform {
...
+                await this.target_file.writev(this.fs_account_config, buffers_to_write);
+                // Hold the ref on the buffers from the JS side
+                this._total_num_buffers += buffers_to_write.length;
+            }
+            if (callback) {
+                if (this.MD5Async) this.digest = (await this.MD5Async.digest()).toString('hex');
+                return callback();
+            }
+        } catch (error) {
+            return callback(error);
```
Comment (guymguym) [inline src/util/chunk_fs.js:73]: Perhaps this should be printed here as warning/error? sounds like this is wouldn't be a good thing that we fail during this process of flushing data from memory to file... same for transform..
```suggestion
            console.error('ChunkFS FLUSH TO FILE FAILED', this.q_size, this._total_num_buffers, error);
            return callback(error);
```

File: src/util/chunk_fs.js [hunk 3]
```diff
@@ -28,54 +28,50 @@ class ChunkFS extends stream.Transform {
...
+                const available_size = config.NSFS_BUF_SIZE - this.q_size;
+                const buf = (available_size < chunk.length) ? chunk.slice(0, available_size) : chunk;
+                this.q_buffers.push(buf);
+                this.q_size += buf.length;
+                if (this.q_size === config.NSFS_BUF_SIZE) await this._flush_buffers();
+                chunk = (available_size < chunk.length) ? chunk.slice(available_size) : null;
+            }
+            return callback();
         } catch (error) {
             return callback(error);
```
Comment (guymguym) [inline src/util/chunk_fs.js:50]:
```suggestion
            console.error('ChunkFS TRANSFORM FAILED', this.q_size, this._total_num_buffers, error);
            return callback(error);
```

File: src/util/chunk_fs.js [hunk 4]
```diff
@@ -28,54 +28,50 @@ class ChunkFS extends stream.Transform {
         this._total_num_buffers = 0;
     }
 
-    _transform(chunk, encoding, callback) {
+    async _transform(chunk, encoding, callback) {
         try {
-            this._process_chunk(chunk, callback);
+            if (this.MD5Async) await this.MD5Async.update(chunk);
+            stats_collector.instance(this.rpc_client).update_namespace_write_stats({
```
Comment (guymguym) [inline src/util/chunk_fs.js:34]: I would add this condition to help running this class it in tests that do not care about stats:
```suggestion
            if (this.rpc_client) {
                stats_collector.instance(this.rpc_client).update_namespace_write_stats({
```

Comment (jeniawhite) [reply src/util/chunk_fs.js:34]: I'm using a dummy RPC for that matter, you want me to revert the dummy RCP code from the unit_tests?
```
const DUMMY_RPC = {
    object: {
        update_endpoint_stats: (...params) => null
    }
};
```

**Comment by guymguym** [comment]
@jeniawhite Can you provide a detailed explanation of what caused the unsettled promise in the code before?
I mainly want to learn from this - and from the info in the PR I couldn't figure it out.
Thanks

**Comment by jeniawhite** [comment]
@guymguym 
There was a problem with ```writev``` writing to invalid addresses of memory
This caused an unhandled promise rejection from the pipeline which caused unsettled promises (the pipeline itself is being awaited for), which in turn caused a leak in the semaphore of the I/O of NSFS
I've wrapped things with try/catch and rewrote the methods to async at the top level (transform/flush)

**Comment by guymguym** [comment]
Thanks.
The writev fix was in #6852 right?
So this fix is required as safety only?

**Comment by jeniawhite** [comment]
@guymguym 
Yes, you are correct
But I need to clarify that this fix is not a safety only
Even with #6852 applied if we would have had a different problem from the native layer or any other exception inside the pipeline's ```transform/flush``` we would have a leak without this code
So this handles any future exceptions from the pipeline and cause them not to create a leak in the semaphore

**Review by guymguym — APPROVED** [review]

**Pull request merged**